### DISCUSSION TITLE
DiaNN-1.8 performance on Linux

### DISCUSSION THREAD

**Discussion by MatteoLacki** [vdemichev/DiaNN#135] [General]
Hello Vadim,

I have made a series of tests on DiaNN on Linux.
In particular, I have been running a script like this:

```
ARGS="--verbose 1 --qvalue 0.01 --matrices --gen-spec-lib --predictor --fasta $FASTAPATH --fasta-search --min-fr-mz 200 --max-fr-mz 1800 --met-excision --cut K*,R*,!*P --missed-cleavages 1 --min-pep-len 7 --max-pep-len 30 --min-pr-mz 300 --max-pr-mz 1800 --min-pr-charge 1 --max-pr-charge 4 --unimod4 --reanalyse --smart-profiling --pg-level 2 --species-genes"

for THREADS in 16 32 64 96 128
do
RES=$RESULTS/threads_$THREADS
mkdir -p $RES
$DIANN $DATAFOLDERS $ARGS --threads $THREADS --out $RES/demo.report.tsv --out-lib $RES/demo_lib.tsv
```
(with other arguments defined elsewhere, the script works, thanks to you).

I was wondering how DiaNN works with different number of assigned threads.
Also, I wanted to check if it is IO bound by putting the data once on HDD and once on a ramdisk.
Here are the outcomes:

![image](https://user-images.githubusercontent.com/3964160/124895146-95906700-dfdc-11eb-9c2a-0a4ce12a591e.png)

As much as I can get it that DiaNN ain't IO bound, I wanted to ask you opinion on the runtimes rising with the number of threads.
I guess, it's not entirely parallizeable and wanted to ask if I do something wrong, or it is general the case.
In any case it ain't a problem: I will gladly run 4 instances of DiaNN simultaneously on different problems.

Best wishes,

**Comment by vdemichev** [comment]
Hi Matteo,

Thank you for testing!
Most likely it's getting limited either by memory bandwidth or by memory allocate/release operations. Thus does not improve when going beyond 32 threads. Please note that, as I understand, this is purely a benchmark of the deep learning predictor, the core DIA-NN code might scale differently, and also differently depending on the nature of the data. And deep learning predictions can usually be reused (they are saved as a .predicted.speclib file).

Yes, running multiple instances simultaneously is likely to always yield higher total throughput, provided the PC has enough RAM.

Best,
Vadim
